src/Microsoft.Tye.Hosting/IngressService.cs

```diff
@@ -0,0 +1,287 @@
+
+                        var uris = new List<Uri>();
+
+                        // For each of the target service replicas, get the base URL
+                        // based on the replica port
+                        for (int i = 0; i < targetServiceDescription.Replicas; i++)
+                        {
+                            foreach (var binding in targetServiceDescription.Bindings)
+                            {
+                                if (binding.Port == null)
+                                {
+                                    continue;
+                                }
+
+                                var port = target.PortMap[binding.Port.Value][i];
+                                var url = $"{binding.Protocol ?? "http"}://localhost:{port}";
+                                uris.Add(new Uri(url));
+                            }
+                        }
+
+                        // The only load balancing strategy here is round robin
+                        long count = 0;
+                        RequestDelegate del = context =>
+                        {
+                            var next = (int)(Interlocked.Increment(ref count) % uris.Count);
+
+                            var uri = new UriBuilder(uris[next])
+                            {
+                                Path = (string)context.Request.RouteValues["path"]
+                            };
+
+                            return context.ProxyRequest(invoker, uri.Uri);
+                        };
+
+                        IEndpointConventionBuilder conventions = null!;
+
+                        if (rule.Path != null)
+                        {
+                            conventions = ((IEndpointRouteBuilder)webApp).Map(rule.Path.TrimEnd('/') + "/{**path}", del);
+                        }
+                        else
+                        {
+                            conventions = webApp.MapFallback(del);
+                        }
+
+                        if (rule.Host != null)
+                        {
+                            conventions.WithMetadata(new IngressHostMetadata(rule.Host));
+                        }
+
+                        conventions.WithDisplayName(rule.Service);
+                    }
+                }
+            }
+
+            foreach (var app in _webApplications)
+            {
+                await app.StartAsync();
+            }
+        }
+
+        public async Task StopAsync(Model.Application application)
+        {
+            foreach (var webApp in _webApplications)
+            {
+                try
+                {
+                    await webApp.StopAsync();
+                }
+                catch (OperationCanceledException)
+                {
+
+                }
+                finally
+                {
+                    webApp.Dispose();
+                }
+            }
+        }
+
+        private class ConnectionRetryHandler : DelegatingHandler
```

Comment:
Put this in a separate class and share with test code. https://github.com/dotnet/tye/blob/master/test/E2ETest/RetryHandler.cs already exists.

Reply:
I can delete the test code but these are a bit different.